dfdaemon/seed/downloader.go

```diff
@@ -0,0 +1,130 @@
+/*
+ * Copyright The Dragonfly Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package seed
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"net/http"
+	"time"
+
+	"github.com/dragonflyoss/Dragonfly/dfget/config"
+	"github.com/dragonflyoss/Dragonfly/pkg/errortypes"
+	"github.com/dragonflyoss/Dragonfly/pkg/httputils"
+	"github.com/dragonflyoss/Dragonfly/pkg/limitreader"
+	"github.com/dragonflyoss/Dragonfly/pkg/ratelimiter"
+
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+)
+
+// downloader manage the downloading of seed file.
+type downloader interface {
+	DownloadToWriterAt(ctx context.Context, rangeStruct httputils.RangeStruct, timeout time.Duration, writeOff int64, writerAt io.WriterAt, rateLimit bool) (length int64, err error)
```

Comment:
The parameter list doesn't contain `url` and it's initialized in creating instance of a concrete implementation. Why is it designed this way?

And this method has 2 things to do:
1. downloads data from remote server.
2. writes data into target writer.

I think that the step 2 may be not the responsibility of downloader interface. Will it be better to remove it from the interface?

Reply:
1. a downloader binds to a seed, so the information of seed will be transferred  by creating instance of a concrete implementation.  In the other word, we may support other protocol like http/https/oss/hdfs ,so the parameters may not be easy to define.
2. in our case, we want to cover the error handler of download. And the data finally should be write ,so we do write in  the interface. And some optimization could be make in the interface, such as buffer Write-through which reduces the costs of copying buffer.